projects/dpg/product_fe_space.h

```diff
@@ -0,0 +1,398 @@
+#ifndef PROJECTS_DPG_PRODUCT_FE_SPACE_H
+#define PROJECTS_DPG_PRODUCT_FE_SPACE_H
+
+/**
+ * @file
+ * @brief Data structure describing (cartesian)-product finite element spaces
+ * @author Philippe Peter
+ * @date June 2019
+ * @copyright MIT License
+ */
+
+#include <vector>
+
+#include <lf/base/base.h>
+#include <lf/uscalfe/uscalfe.h>
+
+#include "dpg.h"
+#include "product_dofhandler.h"
+
+namespace projects::dpg {
+/**
+ * @headerfile projects/dpg/product_fe_space.h
+ * @brief cartesian/prodcut space of finite element functions on a hybrid 2D
+ * mesh.
+ *
+ * @tparam SCALAR underlying scalar type for all components, usually 'double'
+ *
+ *
+ * This class provides functionality similar to the class
+ *lf::uscalfe::UniformScalarFESpace for product/cartesian finite element spaces.
+ *These spaces are of the form
+ *
+ *\f[ U = U_0 \times U_1 \times
+ * \dots \times U_{n-1} \f] where each of the \f$ U_i \f$ is a  function space
+ * of scalar  valued functions (see the description in
+ *ProductUniformFEDofHandler).
+ *
+ * This class weakens some constraints compared to the
+ *lf::uscalfe::UniformScalarFESpace to represent function spaces as they  occur
+ *in the description of DPG methods:
+ *
+ * - There may be multiple global shape functions associated with the same
+ *vertex (usually they will belong to different components)
+ * - Some components may not specify any local shape functions on triangles or
+ *quadrilaterals, but only on segments.
```

Comment:
Yes, in theory this would be possible using a `UniformScalarFESpace`. However the current implementation of the `UniformScalarFESpace` does not allow assigning multiple shape functions to one vertex. This is checked in lines 200-200 of the init method. I will adapt this comment to also focus a bit more on the difference which you pointed out, which is indeed conceptually much more important.

Reply:
Yup, I thought that the doxygen comment coudl be a bit more descriptive ;)

Reply:
Well, `UniformScalarFESpace` is really meant for **scalar and continuous** finite element functions, which always must have a single value at every node. It would be a breach of logic to allow multiple dofs at nodes. If this is required, other types have to be created.

Reply:
I wll take care of updating the documentation of `UniformScalarFESpace`.